stdlib/itertools.pyi

```diff
@@ -101,10 +101,10 @@ class islice(Iterator[_T]):
     def __iter__(self) -> Self: ...
     def __next__(self) -> _T: ...
 
-class starmap(Iterator[_T]):
-    def __init__(self, __function: Callable[..., _T], __iterable: Iterable[Iterable[Any]]) -> None: ...
+class starmap(Iterator[_T_co]):
+    def __new__(cls, __function: Callable[..., _T], __iterable: Iterable[Iterable[Any]]) -> starmap[_T]: ...
     def __iter__(self) -> Self: ...
-    def __next__(self) -> _T: ...
+    def __next__(self) -> _T_co: ...
 
 class takewhile(Iterator[_T]):
```

Comment:
While you're here, maybe change this one too?

Reply:
islice and filterfalse above also look like they should be covariant.

Reply:
Most of them should probably be covariant. My plan is to do all the remaining ones after I've finished adding all the recipes as test cases, so that we have comprehensive regression tests and can be sure we're not breaking anything :-)

Reply:
Here's the full set of changes I think should be made, FWIW. It ends up being a reasonably large diff, hence why I'd prefer to defer it to another PR and keep this one focussed on the bugfix that has a regression test:

<details>
<summary>Full diff</summary>

```diff
--- a/stdlib/itertools.pyi
+++ b/stdlib/itertools.pyi
@@ -34,35 +34,32 @@ class count(Iterator[_N]):
     def __next__(self) -> _N: ...
     def __iter__(self) -> Self: ...

-class cycle(Iterator[_T]):
-    def __init__(self, __iterable: Iterable[_T]) -> None: ...
-    def __next__(self) -> _T: ...
+class cycle(Iterator[_T_co]):
+    def __new__(cls, __iterable: Iterable[_T]) -> cycle[_T]: ...
+    def __next__(self) -> _T_co: ...
     def __iter__(self) -> Self: ...

-class repeat(Iterator[_T]):
-    @overload
-    def __init__(self, object: _T) -> None: ...
-    @overload
-    def __init__(self, object: _T, times: int) -> None: ...
-    def __next__(self) -> _T: ...
+class repeat(Iterator[_T_co]):
+    def __new__(cls, object: _T, times: int = ...) -> repeat[_T]: ...
+    def __next__(self) -> _T_co: ...
     def __iter__(self) -> Self: ...
     def __length_hint__(self) -> int: ...

-class accumulate(Iterator[_T]):
+class accumulate(Iterator[_T_co]):
     if sys.version_info >= (3, 8):
         @overload
-        def __init__(self, iterable: Iterable[_T], func: None = None, *, initial: _T | None = ...) -> None: ...
+        def __new__(cls, iterable: Iterable[_T], func: None = None, *, initial: _T | None = ...) -> accumulate[_T]: ...
         @overload
-        def __init__(self, iterable: Iterable[_S], func: Callable[[_T, _S], _T], *, initial: _T | None = ...) -> None: ...
+        def __new__(cls, iterable: Iterable[_S], func: Callable[[_T, _S], _T], *, initial: _T | None = ...) -> accumulate[_T]: ...
     else:
-        def __init__(self, iterable: Iterable[_T], func: Callable[[_T, _T], _T] | None = ...) -> None: ...
+        def __new__(self, iterable: Iterable[_T], func: Callable[[_T, _T], _T] | None = ...) -> accumulate[_T]: ...

     def __iter__(self) -> Self: ...
-    def __next__(self) -> _T: ...
+    def __next__(self) -> _T_co: ...

-class chain(Iterator[_T]):
-    def __init__(self, *iterables: Iterable[_T]) -> None: ...
-    def __next__(self) -> _T: ...
+class chain(Iterator[_T_co]):
+    def __new__(cls, *iterables: Iterable[_T]) -> chain[_T]: ...
+    def __next__(self) -> _T_co: ...
     def __iter__(self) -> Self: ...
     @classmethod
     # We use type[Any] and not type[_S] to not lose the type inference from __iterable
@@ -70,20 +67,20 @@ class chain(Iterator[_T]):
     if sys.version_info >= (3, 9):
         def __class_getitem__(cls, __item: Any) -> GenericAlias: ...

-class compress(Iterator[_T]):
-    def __init__(self, data: Iterable[_T], selectors: Iterable[Any]) -> None: ...
+class compress(Iterator[_T_co]):
+    def __new__(cls, data: Iterable[_T], selectors: Iterable[Any]) -> compress[_T]: ...
     def __iter__(self) -> Self: ...
-    def __next__(self) -> _T: ...
+    def __next__(self) -> _T_co: ...

-class dropwhile(Iterator[_T]):
-    def __init__(self, __predicate: _Predicate[_T], __iterable: Iterable[_T]) -> None: ...
+class dropwhile(Iterator[_T_co]):
+    def __new__(cls, __predicate: _Predicate[_T], __iterable: Iterable[_T]) -> dropwhile[_T]: ...
     def __iter__(self) -> Self: ...
-    def __next__(self) -> _T: ...
+    def __next__(self) -> _T_co: ...

-class filterfalse(Iterator[_T]):
-    def __init__(self, __predicate: _Predicate[_T] | None, __iterable: Iterable[_T]) -> None: ...
+class filterfalse(Iterator[_T_co]):
+    def __new__(cls, __predicate: _Predicate[_T] | None, __iterable: Iterable[_T]) -> filterfalse[_T]: ...
     def __iter__(self) -> Self: ...
-    def __next__(self) -> _T: ...
+    def __next__(self) -> _T_co: ...

 class groupby(Iterator[tuple[_T_co, Iterator[_S_co]]], Generic[_T_co, _S_co]):
     @overload
@@ -93,23 +90,25 @@ class groupby(Iterator[tuple[_T_co, Iterator[_S_co]]], Generic[_T_co, _S_co]):
     def __iter__(self) -> Self: ...
     def __next__(self) -> tuple[_T_co, Iterator[_S_co]]: ...

-class islice(Iterator[_T]):
+class islice(Iterator[_T_co]):
     @overload
 class groupby(Iterator[tuple[_T_co, Iterator[_S_co]]], Generic[_T_co, _S_co]):
     @overload
@@ -93,23 +90,25 @@ class groupby(Iterator[tuple[_T_co, Iterator[_S_co]]], Generic[_T_co, _S_co]):
     def __iter__(self) -> Self: ...
     def __next__(self) -> tuple[_T_co, Iterator[_S_co]]: ...

-class islice(Iterator[_T]):
+class islice(Iterator[_T_co]):
     @overload
-    def __init__(self, __iterable: Iterable[_T], __stop: int | None) -> None: ...
+    def __new__(cls, __iterable: Iterable[_T], __stop: int | None) -> islice[_T]: ...
     @overload
-    def __init__(self, __iterable: Iterable[_T], __start: int | None, __stop: int | None, __step: int | None = ...) -> None: ...
+    def __new__(
+        cls, __iterable: Iterable[_T], __start: int | None, __stop: int | None, __step: int | None = ...
+    ) -> islice[_T]: ...
     def __iter__(self) -> Self: ...
-    def __next__(self) -> _T: ...
+    def __next__(self) -> _T_co: ...

 class starmap(Iterator[_T_co]):
     def __new__(cls, __function: Callable[..., _T], __iterable: Iterable[Iterable[Any]]) -> starmap[_T]: ...
     def __iter__(self) -> Self: ...
     def __next__(self) -> _T_co: ...

-class takewhile(Iterator[_T]):
-    def __init__(self, __predicate: _Predicate[_T], __iterable: Iterable[_T]) -> None: ...
+class takewhile(Iterator[_T_co]):
+    def __new__(cls, __predicate: _Predicate[_T], __iterable: Iterable[_T]) -> takewhile[_T]: ...
     def __iter__(self) -> Self: ...
-    def __next__(self) -> _T: ...
+    def __next__(self) -> _T_co: ...

 def tee(__iterable: Iterable[_T], __n: int = 2) -> tuple[Iterator[_T], ...]: ...

@@ -270,10 +269,10 @@ class combinations(Iterator[_T_co]):
     def __iter__(self) -> Self: ...
     def __next__(self) -> _T_co: ...

-class combinations_with_replacement(Iterator[tuple[_T, ...]], Generic[_T]):
-    def __init__(self, iterable: Iterable[_T], r: int) -> None: ...
+class combinations_with_replacement(Iterator[tuple[_T_co, ...]]):
+    def __new__(cls, iterable: Iterable[_T], r: int) -> combinations_with_replacement[_T]: ...
     def __iter__(self) -> Self: ...
-    def __next__(self) -> tuple[_T, ...]: ...
+    def __next__(self) -> tuple[_T_co, ...]: ...
```

</diff>